src/ContextualLogger.ts

```diff
@@ -0,0 +1,77 @@
+import { IContext, IJSLoggerDefaultsType, ILogger, ILoggerOpts, ILogLevel, JSLoggerExportType, LogHandler } from "./datatypes";
+export * from "./datatypes";
+import { JSLoggerDefaults as Logger } from "./JSLoggerDefaults";
+import { bind, defineLogLevel } from "./utils";
+
+// Inner class which performs the bulk of the work; ContextualLogger instances can be configured independently
+// of each other.
+export class ContextualLogger implements ILogger {
+  public readonly context: IContext;
+
+  public constructor(defaultContext: IContext) {
+    this.context = defaultContext;
+    this.setLevel(defaultContext.level);
+  }
+
+  // Changes the current logging level for the logging instance.
+  public setLevel(newLevel: ILogLevel) {
+    // Ensure the supplied Level object looks valid.
+    if (newLevel && "value" in newLevel) {
```

Comment:
Not sure we need this runtime check now that we are using types.

Reply:
Yep, that would be obsolete by now :) Neither `null` / `undefined` nor a non-matching interface could be passed here from strict TypeScript code.